Regression tests

schunkes opened the pull request:
# Description

This PR introduces a framework to run complex render tests and compare them to reference results.

The `test_tools.py` module, contains methods to evaluate a test based on the RMSE and a chi2 test.
Especially the chi2 test function needs to be reviewed, since it is adapted from mitsuba2.
(`mitsuba.python.chi2`)

There can also be discussion about which metadata should be included in the [0][1] subplot in the plotting method.

[This](https://en.wikipedia.org/wiki/Pearson%27s_chi-squared_test) wikipedia article defines the chi-squared test as it is used here

# Checklist

- [x] The code follows the relevant coding guidelines
- [x] The code generates no new warnings
- [x] The code is appropriately documented
- [x] The code is tested to prove its function
- [x] The feature branch is rebased on the current state of the `main` branch
- [x] I give permission that the Eradiate project may redistribute my contributions under the terms of its license

leroyvn commented on src/eradiate/test_tools.py:
```diff
@@ -71,3 +75,142 @@ def missing_artefact(filename: PathLike) -> None:
+    return rmse <= rmse_threshold, rmse
+
+
+def assert_chi2(reference, result, significance=0.05):
```
Maybe sort the arguments consistently with `assert_rmse()`?

leroyvn commented on src/eradiate/test_tools.py:
```diff
@@ -71,3 +75,142 @@ def missing_artefact(filename: PathLike) -> None:
+    return passed, p_value
+
+
+def create_archive_files(path: PathLike, filename: str, data: xr.Dataset):
```
I generally find names such as `create_foo()` to be not very informative. But anyway, all this function does is call `to_netcdf()`, which, IMO, isn't really worth a dedicated function. The dir creation can be turned into a one-liner thanks to the `exist_ok` kwarg of [`os.makedirs()`](https://docs.python.org/3/library/os.html#os.makedirs).

leroyvn commented on src/eradiate/test_tools.py:
```diff
@@ -71,3 +75,142 @@ def missing_artefact(filename: PathLike) -> None:
+    data.to_netcdf(os.path.join(path, filename))
+
+
+def create_regression_test_plots(data: xr.Dataset, filename: PathLike, metric: tuple):
```
As previously: maybe something more to-the-point, like (I'll also take the opportunity to fix type hints)
```suggestion
def regression_test_plots(data: xr.Dataset, filename: PathLike, metric: t.Tuple[str, float]):
```

leroyvn commented on src/eradiate/test_tools.py:
```diff
@@ -71,3 +75,142 @@ def missing_artefact(filename: PathLike) -> None:
+    metric : tuple
+        A tuple of the form (metric name, value) to be added to the plots.
+    """
+    eradiate.plot.set_style()
```
This is done by `conftest.py` and you don't want to force it at this level.
```suggestion
```

leroyvn commented on src/eradiate/test_tools.py:
```diff
@@ -71,3 +75,142 @@ def missing_artefact(filename: PathLike) -> None:
+    )
+
+    plt.tight_layout()
+    plt.savefig(filename)
```
Don't forget to close that figure:
```suggestion
    plt.savefig(filename)
    plt.close(fig)
```

leroyvn reviewed (commented):
Hi @schunkes, I don't have much to say about your implementation of Pearson's chi squared test, it seems fine to me. You'll have to document it properly so that readers understand well that you're testing for goodness of fit.

I made a couple of small comments, more on implementation details. On the design side, and with the prospect of having a somewhat uniform interface for all those tests, I think having different return types for each `assert_*()` function can be impractical. Instead, turning those functions into small objects with a log buffer (containing messages and reported metric as a string, just like Mitsuba's `ChiSquaredTest` class) would likely be cleaner and allow for swapping tests more easily . For instance, you could have something like
```python
# Some test code
...
test = RMSETest(values, reference, threshold=1e-3)
assert test.run()
```
and `run()` would output log contents in case of failure.

If you happen to decide that you actually want to run a different kind of test, all you have to do is change the test line—the reporting will follow.

schunkes commented on tests/test_eradiate/regression/romc/test_het04.py:
```diff
@@ -0,0 +1,97 @@
+    reference_path = data_store.fetch("tests/canopies/HET04/het04_brfpp_ref.nc")
+    reference = xr.load_dataset(reference_path)
+
+    passed, rmse = assert_rmse(result.brf.values, reference.brf.values)
```
I can't do that because I use the rmse value inside the archive and plot generation routine.

leroyvn commented on tests/test_eradiate/regression/romc/test_het01.py:
```diff
@@ -0,0 +1,88 @@
+    reference = xr.load_dataset(reference_path)
+    archive_path = metadata.get("archive_path", None)
+
+    if archive_path:
+        datestring = session_timestamp.strftime(format="%Y%m%d")
+        timestring = session_timestamp.strftime(format="%H%M%S")
+        archive_filename = f"{datestring}-{timestring}-het01.nc"
+        archive_filename = os.path.join(archive_path, archive_filename)
+    else:
+        archive_filename = None
```
```suggestion
archive_filename = os.path.join(
    archive_path, f"{session_timestamp:%Y%m%d-%H%M%S}-het01.nc"
) if archive_path else None
```
(to be confirmed)

See also

- https://www.python.org/dev/peps/pep-3101/#controlling-formatting-on-a-per-type-basis
- https://docs.python.org/3/library/datetime.html#datetime.date.__format__

leroyvn commented on src/eradiate/test_tools/regression.py:
```diff
@@ -0,0 +1,232 @@
+import matplotlib.pyplot as plt
+import xarray as xr
+
+from eradiate.typing import PathLike
```
Please use relative imports in package code.

leroyvn commented on tests/test_eradiate/conftest.py:
```diff
@@ -59,3 +59,10 @@ def ert_seed_state():
     return SeedState(0)
+
+
+@pytest.fixture(scope="session")
+def session_timestamp():
+    from datetime import datetime
+
+    return datetime.now()
```
Maybe this should go one level up? Or do you want this fixture to be restricted to non-plugin tests?

leroyvn reviewed (changes requested):
Hi @schunkes, this is taking a nice shape. Some general feedback hereafter.

- Thanks for scoping regression testing-oriented components in a module.
- Your classes should be written with attrs (the rest of our codebase assumes so, and we have infrastructure to facilitate their documentation).
- Your test classes should inherit from a parent interface. At this point, it's just a simple `Test` ABC, with a single `run()` abstract method.
- Old code should be deleted. Much of the docs can be transferred to class field docs.
- Docstrings need some polishing.

leroyvn commented on src/eradiate/test_tools/regression.py:
```diff
@@ -0,0 +1,160 @@
+from ..typing import PathLike
+
+
+def regression_test_plots(data: xr.Dataset, filename: PathLike, metric: tuple):
```
```suggestion
def regression_test_plots(data: xr.Dataset, filename: PathLike, metric: tuple) -> None:
```

leroyvn reviewed (changes requested):
Hi @schunkes, we're almost there. I made a pass on the `regression` module and made specific comments, mostly related with documentation. I also see duplicate code between the two concrete test classes which maybe could be pulled up to the `RenderTest` class; we'll see about that later.

I think the reference data should be gathered in a dedicated directory of the `data`submodule rather than thematically sorted. This kind of storage is more flexible: multiple tests may use the same data, and tests may use data coming from different sources.

I trust there is a logic associating test names, artefact names and reference file names? Is there automation for this yet? This is probably something we'll have to think about implementing in a not-too-distant future.

schunkes commented:
About the data submodule: I'll update the structure tomorrow.

About file names: At the moment the pattern is rather loose and now that you mentioned it, it does actually not work too well for the archive files that are placed in a single output directory. I'll think about this and update that as well.

schunkes commented:
Alright, I reorganized the data submodule and updated the tests accordingly. Additionally I renamed the folders in the data repo to be all lower case, for consistency.

leroyvn commented:
Alright, that seems good now. I'll take a few moments to pimp the branch and PR a little, then I'll merge this. Thanks @schunkes!

leroyvn commented:
Let's merge this. It's missing a developer manual entry but we can add this a little later.
